internal/pkg/merge/merge.go

```diff
@@ -0,0 +1,137 @@
+// ------------------------------------------------------------------------
+// SPDX-FileCopyrightText: Copyright © 2024 bomctl a Series of LF Projects, LLC
+// SPDX-FileName: internal/pkg/merge/merge.go
+// SPDX-FileType: SOURCE
+// SPDX-License-Identifier: Apache-2.0
+// ------------------------------------------------------------------------
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------
+package merge
+
+import (
+	"fmt"
+	"slices"
+
+	"github.com/google/uuid"
+	"github.com/protobom/protobom/pkg/sbom"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/bomctl/bomctl/internal/pkg/db"
+	"github.com/bomctl/bomctl/internal/pkg/options"
+)
+
+func Merge(documentIDs []string, opts *options.MergeOptions) (string, error) {
+	backend, err := db.BackendFromContext(opts.Context())
+	if err != nil {
+		return "", fmt.Errorf("%w", err)
+	}
+
+	backend.Logger.Info("Merging documents", "documentIDs", documentIDs)
+
+	documents, err := backend.GetDocumentsByID(documentIDs...)
+	if err != nil {
+		return "", fmt.Errorf("%w", err)
+	}
+
+	// Make document list a map so it can sort by the ids proved
+	documentMap := make(map[string]*sbom.Document)
+	for _, doc := range documents {
+		documentMap[doc.Metadata.Id] = doc
+	}
+
+	merged, err := performTopLevelMerge(documentIDs, documentMap, opts)
+	if err != nil {
+		backend.Logger.Fatal(err)
+	}
+
+	err = mergeRootNodes(merged)
+	if err != nil {
+		backend.Logger.Fatal(err)
+	}
+
+	backend.Logger.Info("Adding merged document", "sbomID", merged.Metadata.Id)
+
+	err = backend.AddDocument(merged)
+	if err != nil {
+		backend.Logger.Fatal(err)
+	}
+
+	return merged.Metadata.Id, nil
+}
+
+func performTopLevelMerge(sbomIDs []string, documentMap map[string]*sbom.Document,
+	opts *options.MergeOptions,
+) (*sbom.Document, error) {
+	newDocumentID := uuid.New().URN()
+
+	var err error
+
+	merged := sbom.NewDocument()
+
+	if opts.DocumentName != "" {
+		merged.Metadata.Name = opts.DocumentName
+	}
+
+	merged.Metadata.Id = newDocumentID
+	merged.Metadata.Date = timestamppb.Now()
+	merged.Metadata.Version = "1"
+
+	for _, sbomID := range sbomIDs {
+		// Protobom performs add/update on all nodes in list and re-points edges to nodelist
+		err = NewMerger(merged.NodeList).MergeProtoMessage(documentMap[sbomID].NodeList)
+		if err != nil {
+			return nil, fmt.Errorf("failed to merge nodelist: %w", err)
```

Comment:
Should this be logged?

Reply:
You're asking logged vs error?

Reply:
I think I finally traced the whole thing and see how the errors are getting handled. This is fine 👍